### PR TITLE
Use datum arithmetic scalar value

### PR DESCRIPTION
**Pull request by tustvold**
## Which issue does this PR close?

Part of #7353 

## Rationale for this change

This avoids duplicated arithmetic logic, and moves us closer to replacing the custom ScalarValue logic with upstream kernels (#7353)

## Are there any user-facing changes?

The upstream kernels have a couple of differences compared to the existing impls:

* Null + v = Null, whereas previously it was v
* Decimal precision is correctly updated
* Heterogeneous interval arithmetic is not supported, this is instead the responsibility of the type coercion machinery

### CODE REVIEW

File: datafusion/common/src/scalar.rs [hunk 1]
```diff
@@ -2072,53 +1188,48 @@ impl ScalarValue {
         }
     }
 
+    /// Wrapping addition of `ScalarValue`
+    ///
+    /// NB: operating on `ScalarValue` directly is not efficient, performance sensitive code
+    /// should instead make use of vectorized array kernels
     pub fn add<T: Borrow<ScalarValue>>(&self, other: T) -> Result<ScalarValue> {
```
Comment (tustvold) [inline datafusion/common/src/scalar.rs:1195]: The performance of this is likely worse, however, following #7358 the only use of this logic is within the range analysis framework, the median accumulator, and Range window functions, none of which I would consider performance critical

File: datafusion/common/src/scalar.rs [hunk 2]
```diff
@@ -2072,53 +1188,48 @@ impl ScalarValue {
         }
     }
 
+    /// Wrapping addition of `ScalarValue`
+    ///
+    /// NB: operating on `ScalarValue` directly is not efficient, performance sensitive code
```
Comment (alamb) [inline datafusion/common/src/scalar.rs:1193]: I like the performance admonition 👍

**Review by alamb — APPROVED** [review]
This is great -- thank you @tustvold . I took the liberty of merging up from main (and tweaking the performance comment, because I couldn't help myself) but I think we should merge this PR once the tests are  passing

cc @berkaysynnada

**Comment by alamb** [comment]
> The upstream kernels have a couple of differences compared to the existing impls:

These are all bug fixes in my mind

**Pull request merged**